### PR TITLE
GRD-658 bamQC workflow modifications to accomodate revised QC approaches

### PR DESCRIPTION
**Pull request by gavin-peng**
1, Allow multiple bams as well as single bam as inputs
2, Parallelizing over multiple bams and intervals (chromosomes) in filter task, for perfomance boost
3, Added mergeFile task can merge input bams as well as metrics results from splited intervals/mutiple bams, merge both bams and (summing up) metrics collected in filter step
4, Added one regression test for multiple bam inputs 
5, The metrics collected and calculated have some decrepencies compared to original version, as can be seen in Jekins log: https://gsici.hpc.oicr.on.ca/job/bamQC_vidarr/job/GRD-658/11/execution/node/14/log/?consoleFull
this may as a result of spliting intervals in filter task.
current passed regression test used new metrics generated by new version for two tests

### CODE REVIEW

File: .bamQC.wdl [hunk 1]
```diff
@@ -0,0 +1,1130 @@
+version 1.0
```
Comment (iainrb) [inline .bamQC.wdl]: What does this file do? Could it have been committed by mistake?

File: README.md [hunk 2]
```diff
@@ -4,17 +4,6 @@ QC metrics for BAM files
 
 ## Overview
```
Comment (iainrb) [inline README.md]: Overview section is now empty -- can remove the header if nothing is there

File: README.md [hunk 3]
```diff
@@ -126,34 +138,282 @@ Parameter|Value|Default|Description
...
-
+`result`|File|json file that contains metrics and meta data described in https://github.com/oicr-gsi/bam-qc-metrics/blob/master/metrics.md
 
-## Niassa + Cromwell
 
-This WDL workflow is wrapped in a Niassa workflow (https://github.com/oicr-gsi/pipedev/tree/master/pipedev-niassa-cromwell-workflow) so that it can used with the Niassa metadata tracking system (https://github.com/oicr-gsi/niassa).
-
-* Building
+## Commands
+ This section lists command(s) run by WORKFLOW workflow
```
Comment (iainrb) [inline README.md]: change WORKFLOW -> bamQC

**Review by iainrb — CHANGES_REQUESTED** [review]
- Should `.bamQC.wdl` be removed?
- Minor suggestions for README
- Files have been made executable, but this is not necessary -- can remove execute permissions in git with `git update-index --chmod=-x FILE1 FILE2 ...` followed by `git commit`
- Otherwise looks good!

File: .bamQC.wdl (hunk 1 above)
Comment (gavin-peng) [reply .bamQC.wdl]: Thanks, my mistake, it somehow sneaked in.

File: README.md (hunk 2 above)
Comment (gavin-peng) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (gavin-peng) [reply README.md]: done, thanks

Comment (gavin-peng) [reply README.md]: corrected

File: README.md (hunk 2 above)
Comment (gavin-peng) [reply README.md]: I added back the overview contents.

**Review by lheisler — APPROVED** [review]
our olives will only use this workflow in two modes
ie. either call-ready with 1 or more bam files, or lane-level with only 1 bam file

but what happens if the workflow is called in a different way 
ie. a different mode argument, or calling lane-level but passing 2 or more bam files

here qcReadyBam won't get a value.  So the workflow will simply fail with an error.
I think this is okay, but wondering if there is a better approach, to have the workflow run complete and report on this.  (but maybe it makes more sense to generate an error, so taht this can be addressed in pipeline should it happen for some reason..)

**Comment by gavin-peng** [comment]
If use different value (other then 'lane_level' and 'call_ready') or more than one bam in lane_level mode, the worlkflow would fail. Since only olive will set this parameter value, the olive shoud use correct values. I think this is a convention in some sense, for example if the "reference" parameter set to values other than "hg38", "hg19", "mm10", the workflow would also fail.

**Comment by iainrb** [comment]
I agree with Gavin's comment -- it's fine if a workflow only works on parameters from a list of valid options. I'm sure we have other examples. Unfortunately WDL doesn't seem to support a Java-style enum to specify the permitted values. But if possible we should make it clear in documention, eg. "Path to human genome FASTA reference" -> "Path to genome FASTA reference (must be hg19, hg38 or mm10)"

**Review by iainrb — APPROVED** [review]

**Comment by lheisler** [comment]
> that make sense, there are many possible ways to give incorrect input, and the workflow failing is an indication of this.  It should not be an issue with the olives in production, and the author should be aware of the required input (or would discover this when testing in stage).

**Comment by lheisler** [comment]
oops, i seemed to have closed the PR by mistake.., reopening for merge etc

**Pull request merged**